### DISCUSSION TITLE
UPS con NUT con Docker su Raspbian - si interrompe collegamento

### DISCUSSION THREAD

**Discussion by poliant** [indomus/forum#26] [GENERALE - Discussioni generiche]
ciao,
ho seguito la guida per integrare un UPS con container docker su raspbian ([qui](https://indomus.it/guide/gestire-un-ups-con-raspberry-pi-tramite-network-ups-tools-nut-con-docker-su-raspbian/)).

Ho un tecnoware 1100 collegato via seriale al RPI4 (/dev/bus/usb/001/010) e ho configurato il container docke come segue:
`environment:`
`  - "PORT=auto"`
`  - "NAME=tecnoware"`
`  - "DESCRIPTION=Technoware Era Plus 1100"`
`  - "DRIVER=blazer_usb"`
`- "API_USER=xxxxxx"`
`network_mode: host`
`privileged: true`
`restart: unless-stopped`

e spesso capita che perdo il collegamento seriale e devo buttare giù il container, cancellarlo e ricrearlo.

Qualcuno sa se posso automatizzare il tutto tipo per fare polling e controllare che il collegamento USB funziona ed in caso non c'è risposta dall'ups, far ripartire il servizio?

grazie

**Comment by indomus** [comment]
Prova correggendo il campo port. Indica esplicitamente la porta, anziché impostare auto

**Reply by poliant** [reply]
stesso problema....

ieri ho provato a spostare il cavo USB dall'HUB estreno collegato al RPI a porta USB diretta su RPI, vediamo se cambia qualcosa

**Reply by poliant** [reply]
non cambia nulla!!!....

e per qualche motivo il nome dell'interfaccia continua a cambiare (/dev/bus/usb/001/010, /dev/bus/usb/001/011 ed ora /dev/bus/usb/001/012)... ho rimesso auto nela PORT

ma se non fermo il container e lo ricreo non vedo il UPS....

**Reply by snakuzzo** [reply]
Prova a dare un occhiata ad `udev` ... dovresti risolvere

**Reply by poliant** [reply]
.... che devo cercare?
praticamente continua a cambiare l'interfaccia e l'unico modo per avere il collegamento è fermare il container e ricrearlo

**Reply by snakuzzo** [reply]
[QUI](https://wiki.debian.org/it/udev) trovi descritto cos'è `udev`

Dopo aver letto l'articolo (e nello specifico il paragrafo intitolato **Nomi di device persistenti**) puoi provare a fare così...

1. Esegui il seguente comando...

 `lsusb`
 
ed otterrai ad esempio questo risultato...

```
Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 002 Device 002: ID 0bda:0316 Realtek Semiconductor Corp. USB3.0-CRW
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 003: ID 13d3:56a6 IMC Networks Integrated Camera
Bus 001 Device 002: ID 8087:0a2b Intel Corp. 
Bus 001 Device 005: ID 10c4:ea60 Silicon Labs CP210x UART Bridge
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```

Come vedi vengono elencate tutte le periferiche USB. Ammettiamo il caso che il tuo UPS sia quello chiamato `Silicon Labs CP210x UART Bridge`

Recuperi i valori di `idVendor` ed `idProduct`...che sarebbero quelli separati dai due punti. Il primo è `idVendor`, il secondo è `idProduct`. Quindi, considerato l'esempio appena scritto, saranno:

```
idVendor: 10c4
idProduct: ea60
```

2. Crei una regola udev per fare in modo che venga assegnata sempre la stessa label (che in realtà poi è un symlink) a quel device con quegli idVendor e idProduct...

```
sudo vi /etc/udev/rules.d/99-usb-serial.rules 
SUBSYSTEM=="tty", ATTRS{idVendor}=="10c4", ATTRS{idProduct}=="ea60", SYMLINK+="myUPS"
```

3. Scollega l'UPS e forza l'esecuzione delle regole udev con il seguente comando...

```
sudo udevadm control --reload-rules && sudo udevadm trigger
```

4. Ricollega il cavo USB ed esegui il seguente comando...

`ls -lart /dev/myUPS`

...e dovresti ottenere un risultato come questo:

`lrwxrwxrwx   1 root     root               7 apr  1 17:06 /dev/myUPS -> ttyUSB0`

5. A questo punto nella configurazione inserisci `/dev/myUPS` ... e così non dovresti più avere problemi.

Fammi sapere se risolvi.

**Reply by poliant** [reply]
grazie Diego del supporto.
Ho  creato la regola seguente e al restart tutto funziona

`SUBSYSTEM=="usb", ATTRS{idVendor}=="0665", ATTRS{idProduct}=="5161", SYMLINK+="myUPS"`

Ho provato a scollegare e ricollegare la USB dell'UPS ed il link simbolico si aggiorna correttamente, MA non riesco a comunicare con il UPS (come si vede sotto)
![image](https://user-images.githubusercontent.com/26714536/113757857-67686600-9713-11eb-8a1b-b28d621841f0.png)

Sono costretto a fermare il container, cancellarlo e ricrearlo

**Reply by snakuzzo** [reply]
Questo comportamento è normalissimo. Funziona proprio così, non puoi fare altrimenti. Devi per forza di cose restartare il container

**Reply by poliant** [reply]
questo è un problema perché quello che mi accade è che ogni tanto la porta USB cambia (forse perché il NUT per qualche motivo perde temporaneamente la connessione USB) e io perdo la connessione senza la possibilità di accorgermene...
Quindi pensavo se c'è una specie di watchdog che in caso in cui si accorge che la connessione è persa riavvia il container in automatico e si ristabilisce la connessione

**Reply by snakuzzo** [reply]
No...dovrebbe essere previsto dal container che hai creato. Puoi provare a modificare l'immagine costruendotene una tua oppure vai su dockerhub o github e cerchi se c'è qualche altra immagine al posto di quella che abbiamo indicato in guida.
Resta il fatto che quella cosa che di punto in bianco ti cambia la porta USB, senza che tu faccia nulla, è un problema

**Reply by poliant** [reply]
si è un bel problema!!!!

Ho provato anche su altro RPi direttamente come applicazione e non come container e il problema permane quindi non credo che ricreando l'immagine possa andare.
Proverò a fare un file abtch che fa il ping del servizio ed in caso ricrea il container. Vediamo come va nei prox giorni

**Reply by snakuzzo** [reply]
Non serve ricreare il container, ma restartarlo.
Ad ogni modo credo che la cosa più giusta sia quella di segnalare la cosa tramite una issue sulla pagina del progetto github di quella immagine, o magari collaborare (scrivendo codice) per fare in modo che l'applicazione faccia un controllo e si ricolleghi...e non che si pianti dovendo restartare il servizio

**Reply by Marcellino78** [reply]
ho il tuo stesso prolema... il nut perde connessione e rinomina la porta a cui è collegato...

**Reply by snakuzzo** [reply]
Sì... per la porta vi ho scritto come risolvere.
Per il discorso del dover restartare il container...sarebbe lo stesso utilizzando quell'applicazione come servizio e non come container docker. Dovete cercare altro se questa cosa non vi sta bene, oppure contribuite allo sviluppo di quell'immagine sul repo github relativo :)

**Reply by poliant** [reply]
non sapendo scrivere codice ;-) ho aperto una issue su github ([qui](https://github.com/instantlinux/docker-tools/issues/67))

vediamo come prosegue

**Reply by Marcellino78** [reply]
non riesco a creare la regola...
mi puoi aiutare!!

**Comment by cobradodo** [comment]
Grazie @snakuzzo 
seguendo le tue dritte ora funziona :-D
E' partito al primo colpo... il link si aggiorna. ora provo a togliere e mettere la USB con il container avviato

**Reply by poliant** [reply]
ciao @cobradodo 
hai fatto la prova di scollegare e ricollegare la USB?
A me, anche se il link si aggiorna correttamente, il container perde il collegamento al UPS....

**Reply by cobradodo** [reply]
Guarda...ieri la stavo facendo e purtroppo mi si è schiantato il raspberry...ora sto lottando per sistemare tutto e proseguirò la prova...

**Reply by Marcellino78** [reply]
si anchi'io stesso problema... mi cade tutto.
non funziona ne zwawe.. ne deconz.. ne nut...

**Reply by snakuzzo** [reply]
Sì ma NUT non c'entra niente col resto

**Reply by cobradodo** [reply]
No direi proprio che non centra nulla.

**Comment by akela-it** [comment]
Ciao a tutti, 
rispondo con più di un anno di ritardo sperando abbiate ormai risolto. Scrivo comunque come sono riuscito a risolvere il problema perché ho letto in giro che tanti hanno avuto lo stesso problema ma nessuno ha indicato una soluzione. 
Come @poliant anche io ho acquistato il tecnoware era plus 1100 indicato nella guida indomus per essere sicuro che tutto funzionasse correttamente ma mi sono trovato nella sua stessa situazione e nonostante i buoni consigli di @snakuzzo il mio Raspberry pi 2 ogni tanto perdeva la connessione con l'UPS (lo so che Rpi2 è un po' vecchio ma questo passa il convento e sono ormai 5 anni che fa il suo sporco lavoro e lo fa anche bene). 
Ci ho perso parecchio tempo pensando che il problema potesse dipendere dalla distribuzione di NUT o che fosse un problema dell'UPS ma alla fine ho scoperto che il problema era legato all'utilizzo troppo "pesante" delle porte USB del Raspberry. In particolare sulle porte USB avevo collegato l'UPS, l'hub Meross e la chiavetta USB wifi per collegare home assistant al wifi di casa (il raspberry pi 2 non ha scheda wifi integrata). 
Ho risolto collegando il raspberry alla rete ethernet ed eliminando l'adattatore wifi, così facendo il problema della disconnessione è sparito e ora l'UPS rimane sempre collegato al NUT. Quindi il mio consiglio è di controllare se oltre all'UPS avete collegato anche altri dispositivi al Raspberry ed eventualmente togliete ciò che potete; la guida indomus, il docker NUT e l'UPS consigliato funzionano perfettamente.

**Reply by snakuzzo** [reply]
io però lo utilizzavo lo stesso su un RPI3B ... avendo tutte e 4 le porte USB occupate da:
- UPS
- Sonoff RF Bridge
- SSD
- ESP866

**Reply by akela-it** [reply]
Mi sono dimenticato di scrivere che alimentavo anche due sonoff 5v1c con una porta USB. Ora collegato al RPI2 ho solo l'UPS e l'hub meross e, senza cambiare null'altro, non ho mai perso collegamento con UPS.

**Comment by poliant** [comment]
grazie per i feedback.
Io ho un RPI4 con boot da SSD quindi su una delle porte vi è il controller dell'SSD che non posso sconnetere... sulle altre USB ho
 - nulla sull'altra USB3.0
 - UPS 
 - HUB USB 2.0 dove ho collegate varie periferiche

Tutto funziona alla perfezione eccetto il UPS che dopo un po' (in maniera del tutto random) che faccio partire il container, si scollega!

A proposito, nessun risposta alla issue che ho aperto in github

**Reply by akela-it** [reply]
Hai provato a disconnettere l'hub usb con le diverse periferiche per qualche giorno? Alla issue su github, alla quale abbiamo dato supporto io e KC-indomus, probabilmente nessuno darà feedback perché, se ho ragione io, non si tratta di un problema di NUT.
A me, dopo la disconnessione definitiva della chiavetta wifi e dei due sonoff 5v1c, l'ups non si è più disconnesso e il problema che avevo era identico al tuo. Se il tuo problema riguarda l'SSD invece la vedo dura. Puoi alimentarlo con una alimentazione esterna al raspberry?

**Reply by poliant** [reply]
eccomi nuovamente. Dopo mille prove, test che non risolvevano il problema qualche giorno fa nel palazzo è stata staccata la corrente per fare dei lavori ed il tecnoware non ha retto il carico (raspi4, modem fibra, router wifi ed HD esterno) e si è spenta praticamente in meno di 10 sec!!!! La faccio breve, mi hanno mandato in garanzia un nuovo tecnoware e da quando l'ho connesso non ho più avuto disconnessioni strane (sono circa 12h che va), vediamo se era un problema "occulto" dell'HW del tecnoware....
